### PR TITLE
Feat/msdk 274 refactor instructions procedures

### PR DESCRIPTION
**Pull request by shuffledex**

### CODE REVIEW

File: src/api/procedures/modifyInstructionAuthorization.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+export interface ModifyInstructionAuthorizationParams {
+  operation: InstructionAuthorizationOperation;
+}
+
+/**
+ * @hidden
+ */
+export type Params = ModifyInstructionAuthorizationParams & {
+  id: BigNumber;
+};
```
Comment (monitz87) [inline src/api/procedures/modifyInstructionAuthorization.ts:22]: No need for two separate types here, they can be merged into a single interface

File: src/api/entities/Instruction/types.ts [hunk 2]
```diff
@@ -47,3 +47,9 @@ export interface InstructionAuthorization {
   identity: Identity;
   authorizationStatus: AuthorizationStatus;
 }
+
+export enum InstructionAuthorizationOperation {
```
Comment (monitz87) [inline src/api/entities/Instruction/types.ts:51]: This should be an internal type

File: src/api/procedures/utils.ts [hunk 3]
```diff
@@ -29,3 +33,51 @@
 //     });
 //   }
 // }
+
+/**
+ * @hidden
+ */
+export async function checkInstructionValidation(
```
Comment (monitz87) [inline src/api/procedures/utils.ts:40]: assertInstructionValid

File: src/api/procedures/modifyInstructionAuthorization.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      }
+
+      this.addTransaction(tx.settlement.authorizeInstruction, {}, rawInstructionId, [
+        rawPortfolioId,
+      ]);
+
+      return instruction;
+    }
+    case InstructionAuthorizationOperation.Unauthorize: {
+      if (authorizationStatus === AuthorizationStatus.Pending) {
```
Comment (monitz87) [inline src/api/procedures/modifyInstructionAuthorization.ts:69]: This error should also be thrown on `AuthorizationStatus.Rejected`

File: src/api/procedures/modifyInstructionAuthorization.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        throw new PolymeshError({
+          code: ErrorCode.ValidationError,
+          message: 'The Instruction is unauthorize',
```
Comment (monitz87) [inline src/api/procedures/modifyInstructionAuthorization.ts:72]: The instruction is not authorized

File: src/api/procedures/modifyInstructionAuthorization.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+ * @hidden
+ */
+export type Params = ModifyInstructionAuthorizationParams & {
+  id: BigNumber;
+};
+
+/**
+ * @hidden
+ */
+export async function prepareModifyInstructionAuthorization(
```
Comment (monitz87) [inline src/api/procedures/modifyInstructionAuthorization.ts:27]: Why are you only returning the instruction in 2/3 cases?

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/procedures/modifyInstructionAuthorization.ts (hunk 6 above)
Comment (shuffledex) [reply src/api/procedures/modifyInstructionAuthorization.ts:27]: I thought if operation is to reject the Instruction, return it seems some weird

Comment (monitz87) [reply src/api/procedures/modifyInstructionAuthorization.ts:22]: Makes sense, but rejecting is not deleting. The instruction can still be re-authorized in some cases and the pertinent data still exists

File: src/api/procedures/modifyInstructionAuthorization.ts [hunk 7]
```diff
@@ -66,10 +61,13 @@ export async function prepareModifyInstructionAuthorization(
       return instruction;
     }
     case InstructionAuthorizationOperation.Unauthorize: {
-      if (authorizationStatus === AuthorizationStatus.Pending) {
+      if (
+        authorizationStatus === AuthorizationStatus.Pending ||
+        authorizationStatus === AuthorizationStatus.Rejected
```
Comment (monitz87) [inline src/api/procedures/modifyInstructionAuthorization.ts:66]:
```suggestion
        [AuthorizationStatus.Pending, AuthorizationStatus.Rejected].includes(authorizationStatus)
```

**Review by monitz87 — CHANGES_REQUESTED** [review]
Return the instruction at the end of the procedure instead of on only 2 of 3 cases

**Review by monitz87 — APPROVED** [review]

**Pull request merged**